forest_benchmarking/classical_logic/ripple_carry_adder.py

```diff
@@ -0,0 +1,367 @@
+
+There are many other classical logic primitives that can be coded see
+e.g.
+
+[VBE96] Quantum networks for elementary arithmetic operations
+        Vedral,  Barenco, Ekert
+        Phys. Rev. A 54, 147 (1996)
+        https://doi.org/10.1103/PhysRevA.54.147
+        https://arxiv.org/abs/quant-ph/9511018
+"""
+from typing import Sequence, Tuple
+import networkx as nx
+import numpy as np
+from numpy import pi
+from scipy.spatial.distance import hamming
+
+from pyquil.gates import CNOT, CCNOT, X, I, H, CZ, MEASURE, RESET
+from pyquil import Program
+from pyquil.api import QuantumComputer
+from pyquil.unitary_tools import all_bitstrings
+
+from forest_benchmarking.readout import _readout_group_parameterized_bitstring
+from forest_benchmarking.classical_logic.primitives import *
+
+
+def assign_registers_to_line_or_cycle(start: int, graph: nx.Graph, num_length: int):
+    """
+    From the start node assign registers as they are laid out in the ideal circuit diagram in
+    [CDKM96].
+
+    Assumes that the there are no dead ends in the graph, and any available neighbor can be
+    selected from the start without any further checks.
+
+    :param start: a node in the graph from which to start the assignment
+    :param graph: a graph with an unambiguous assignment from the start node, e.g. a cycle or line
+    :param num_length: the length of the bitstring representation of one summand
+    :return: the necessary registers and ancilla labels for implementing an adder program to add
+        the numbers a and b. The output can be passed directly to adder()
+    """
+    if 2 * num_length + 2 > nx.number_of_nodes(graph):
+        raise ValueError("There are not enough qubits in the graph to support the computation.")
+
+    graph = graph.copy()
+
+    register_a = []
+    register_b = []
+
+    # set the node at start, and assign the carry_ancilla to this node.
+    node = start
+    carry_ancilla = node
+    neighbors = list(graph.neighbors(node))
+
+    idx = 0
+    while idx < 2 * num_length:
+        # remove the last assigned node to ensure it is not reassigned.
+        last_node = node
+        graph.remove_node(last_node)
+
+        # crawl to an arbitrary neighbor node if possible. If not, the assignment has failed.
+        if len(neighbors) == 0:
+            raise ValueError("Encountered dead end; assignment failed.")
+        node = neighbors[0]
+        neighbors = list(graph.neighbors(node))
+
+        # alternate between assigning nodes to the b register and a register, starting with b
+        if (idx % 2) == 0:
+            register_b.append(node)
+        else:
+            register_a.append(node)
+
+        idx += 1
+    # assign the z_ancilla to a neighbor of the last assignment to a
+    z_ancilla = next(graph.neighbors(node))
+
+    return register_a, register_b, carry_ancilla, z_ancilla
+
+
+def prepare_bitstring(bitstring: Sequence[int], register: Sequence[int], in_x_basis: bool = False):
```

Comment:
I could probably combine this with _readout_group_bitstring from the readout module, but I'm not sure where to put the combination.

Reply:
I just made an issue do do this, see https://github.com/rigetti/forest-benchmarking/issues/35 .